EnumObjectTypes.js

```diff
@@ -1,82 +1,106 @@
 /**
  *
- * Enumerate object types from nt!ObpRootDirectoryObject
+ * Enumerate all objects in nt!ObpRootDirectoryObject
  *
  */
 "use strict";
 
 
 const log = x => host.diagnostics.debugLog(x + "\n");
-
+const getHeader = x => x.address.subtract(host.getModuleType("nt", "_OBJECT_HEADER").fields.Body.offset);
+const getName = x  => host.createTypedObject(getHeader(x), "nt", "_OBJECT_HEADER").ObjectName
+const getTypeName = x  => host.createTypedObject(getHeader(x), "nt", "_OBJECT_HEADER").ObjectType
 
 /**
  *
  */
-function *ListObjects()
+function *DumpDirectory(objectDirectory, parentName)
 {
-
-    var ptrsz = host.namespace.Debugger.State.PseudoRegisters.General.ptrsize;
-    var hdrsz = host.evaluateExpression("sizeof(_OBJECT_HEADER)");
-
     //
-    // We must substract because _OBJECT_HEADER and the object body itself overlap by sizeof(void*)
-    // http://codemachine.com/article_objectheader.html
+    // Create the full directory name
     //
-    var hdroff = hdrsz - ptrsz;
+    var rootName = parentName + "\\" + getName(objectDirectory).slice(1, -1);
+    if (rootName === "\\\\") rootName = "";
 
     //
-    // x nt!ObpRootDirectoryObject
-    //
-    var pObpRootDirectoryObject = host.getModuleSymbolAddress("nt", "ObpRootDirectoryObject");
-
+    // Dump the 37 hash buckets
     //
-    // dx (_OBJECT_DIRECTORY*)&nt!ObpRootDirectoryObject
-    //
-    var RootDirectoryObject = host.createPointerObject(pObpRootDirectoryObject, "nt", "_OBJECT_DIRECTORY *");
-
-
-    //
-    // On each HashBucket
-    //
-    for(var i=0; i<RootDirectoryObject.HashBuckets.Count(); i++)
+    for (var bucketEntry of objectDirectory.HashBuckets)
     {
-        try
+        //
+        // Only if non-empty
+        //
+        if (!bucketEntry.isNull)
```

Comment:
This is more performant and less error-prone than using "try" -- we can simply check for Nullity.